packages/blade/src/components/FileUpload/_decisions/decisions.md

```diff
@@ -0,0 +1,398 @@
+# FileUpload Decisions <!-- omit in toc -->
+
+The `FileUpload` component is used to handle file attachments, including the drag-and-drop interaction. It can be used in both controlled and uncontrolled forms. Primarily, it is used to upload files to a server or to display a list of uploaded files.
+
+This document outlines the API details of the `FileUpload` component, encompassing their structural composition, functional attributes, and visual representation.
+
+<img width="100%" src="./single-upload-variant.png" alt="Single File Upload">
+<img width="100%" src="./multiple-upload-variant.png" alt="Multiple Files Upload">
+<img width="100%" src="./file-upload-item-variants.png" alt="FileUploadItem Variants ">
+
+- [Design](#design)
+- [Anatomy](#anatomy)
+- [`FileUpload` Props](#fileupload-props)
+- [Usage](#usage)
+  - [Submit Button File Upload (Uncontrolled)](#submit-button-file-upload-uncontrolled)
+  - [Direct Upload with Progress Display (Controlled)](#direct-upload-with-progress-display-controlled)
+- [Examples](#examples)
+  - [Uncontrolled Usage](#uncontrolled-usage)
+    - [Single File selection](#single-file-selection)
+    - [Multiple File selection:](#multiple-file-selection)
+  - [Controlled Usage](#controlled-usage)
+    - [Direct Upload with Progress Display](#direct-upload-with-progress-display)
+- [Accessibility](#accessibility)
+
+## Design
+
+[Figma Link](https://www.figma.com/file/jubmQL9Z8V7881ayUD95ps/Blade-DSL?type=design&node-id=78150%3A2625&mode=design&t=bKQvPInRmAoeAAFV-1) to all variants of the `FileUpload` component.
+
+## Anatomy
+
+<img width="100%" src="./file-upload-anatomy.png" alt="Single File Upload">
+<img width="100%" src="./file-upload-item-anatomy.png" alt="Multiple Files Upload">
+
+## `FileUpload` Props
+
+```ts
+// Check the File type from MDN for more details: https://developer.mozilla.org/en-US/docs/Web/API/File
+interface BladeFile extends File {
+  /**
+   * The unique identifier of the file.
+   */
+  id: string;
+  /**
+   * The file's upload status.
+   */
+  status?: 'uploading' | 'success' | 'error';
+  /**
+   * The percentage of file upload completion.
+   */
+  percent?: number;
+  /**
+   * Text indicating an error state
+   */
+  errorText?: string;
+}
+
+type BladeFileList = BladeFile[];
+
+type FileUploadCommonProps = {
```

Comment:
What callback would the user get when an upload is cancelled midway?

Reply:
As per the current API, `onRemove` will be triggered when `CloseIcon` or `TrashIcon` is clicked. But users may want to use different logic so we should introduce a new`onCancel` prop. Otherwise they would need to rely on `file.status` which may not be ideal.

<img width="922" alt="Screenshot 2024-02-20 at 2 44 56 PM" src="https://github.com/razorpay/blade/assets/46647141/74630c72-abca-47cd-9127-99acc6f94c8a">

Reply:
`onRemove` theek hai na?

So upload can get cancelled 2 ways
1. User cancels it - we call `onRemove` because user can cancel it only by clicking the X button
2. Internet/Server Failure - what will we call here?

Reply:
>onRemove theek hai na?

So we trigger `onRemove` when either CloseIcon Or TrashIcon is clicked, right?

Reply:
Thinking more no this, I feel we should have 2 different props `onRemove` & `onCancel` because users may want different behavior when the file is uploading or fails to upload and when it has been successfully uploaded.

Reply:
Wouldn't server failures be handled on user's end wherever they are calling the API? How will we call that onCancel?

```js
try {
  // fetch api call
} catch (err) {
 // this is server failure
}
```

Reply:
```js
try {
  // fetch api call
} catch (err) {
 // file.status = "error"
// file.errorText = error.message
}
```

After this the file is displayed in error state:

<img width="223" alt="Screenshot 2024-02-27 at 11 36 05 AM" src="https://github.com/razorpay/blade/assets/46647141/03ff3711-940c-4554-9844-6232ae9ddd10">

Now clicking the close icon triggers `onCancel`. Also trigger when clicking the close icon when uploading files:
<img width="215" alt="Screenshot 2024-02-27 at 11 37 30 AM" src="https://github.com/razorpay/blade/assets/46647141/34fa0a24-0289-4219-abfa-eb8cc36fc29f">

@saurabhdaware

Reply:
no I mean is it us who call API internally in our component? or do consumers call it?

Reply:
>no I mean is it us who call API internally in our component? or do consumers call it?

@saurabhdaware Consumers have control of the submit logic, check examples I've added in this doc for more info

Reply:
so clicking on close icon always calls `onCancel` right? should we call it `onCloseButtonClick`, `onDismiss` in that case similar to our other components?

Because to me, cancel implies that uploading got cancelled but in case where we show error and show close icon, we're not exactly cancelling the upload request no? upload request just failed the validation and the error was dismissed by consumer 🤔  
<img width="273" alt="image" src="https://github.com/razorpay/blade/assets/30949385/2c0f33ca-11af-4f1c-a98d-37248f60d698">

Reply:
Yeah, `onDismiss` is more suitable.